### DISCUSSION TITLE
Reverse proxy base missing part behind /

### DISCUSSION THREAD

**Discussion by jerfeygilbert** [zazuko/trifid#95] [Q&A]
Hi, 

We use Trifid on our server with reverse proxy for the dereferencing with the base URL ‘[https://labs.kadaster.nl/trifid-kg/’]. Unfortunately the links on this page to other items with the same sparqlEndpointUrl and datasetBaseUrl start with ‘[https://labs.kadaster.nl/’] (without the ‘trifid-kg’). We tried using the ‘X-Forwarded-Host’ parameter for the reverse proxy, but without wanted result. 

(For now we adjusted the function ‘renderLink’ in the render-ls.js, to add the ‘trifid-kg’. But this is not a clean solution and it is only for links)

Is there perhaps an example how other users solved this issue? Thanks in advance for the help!

Example:
[https://labs.kadaster.nl/trifid-kg/kadaster/kg/id/geo-shape/bag-0200100000001088]

**Comment by ktk** [comment]
We had a look at what you did. I'm not sure if we fully understand it, I think it would be easier to have a quick call on what you try to do and where the data is coming from. Would that be ok?

**Reply by jerfeygilbert** [reply]
That would be great. Thanks. Will contact you for date and time.

**Comment by bergos** [comment]
It should be possible to use the rewrite feature. Usually, it is used for developer environments, but it should be worth trying to add the following lines to your config.json:

```javascript
{
  "datasetBaseUrl": "https://labs.kadaster.nl/",
  "handler": {
    "root": {
      "module": "trifid-handler-sparql",
      "path": "/trifid-kg/",
      "options": {
        "resourceExistsQuery": "ASK { <${iri}> ?p ?o }",
        "resourceGraphQuery": "DESCRIBE <${iri}>",
        "containerExistsQuery": "ASK { ?s a ?o. FILTER REGEX(STR(?s), \"^${iri}\") }",
        "containerGraphQuery": "CONSTRUCT { ?s a ?o. } WHERE { ?s a ?o. FILTER REGEX(STR(?s), \"^${iri}\") }"
      }
    }
  }
}
```

`datasetBaseUrl` will be used as a replacement for the base URL. The base URL is generated from the path of the middleware. `handler.root.path` defines the base URL of the middlware.

The rewrite feature is not recommended for production instances. There is a slight performance decrease, and there can be issues with the SPARQL proxy.

**Reply by jerfeygilbert** [reply]
Thanks for the quick response, but it does not (fully) solve our question. Guess it will be easier and quicker to have a quick call as mentioned by Adrian in previous comment (contacted him for possible date time).

**Reply by kad-kolka** [reply]
https://labs.kadaster.nl/trifid-kg is the website where we access the trifid data. The reverse proxy passes https://labs.kadaster.nl/trifid-kg(.*) (so everything after https://labs.kadaster.nl/trifid-kg) to a docker container that runs on your default port (as part of a docker-compose for the website). The dataset starts with https://data.labs.kadaster.nl/(kadaster). The problem that we have now is that some UR(I/L)s point to https://data.labs.kadaster/.... en some to  https://labs.kadaster.nl/.... (and that should be data.labs.kadaster.nl.

**Reply by l00mi** [reply]
@kad-kolka Can you please share your current config and two examples of the behaviour.

If there is some error in the logic of the rewriting in case of multiple subdomains (as suspected) it would most probably stem from here: https://github.com/zazuko/camouflage-rewrite/blob/master/index.js

**Reply by jerfeygilbert** [reply]
Hi @l00mi, 

Thanks for the call! Our configuration:

{
    "baseConfig": "trifid:config-sparql.json", 
    "sparqlEndpointUrl": "https://api.labs.kadaster.nl/datasets/kadaster/kg/services/default/sparql",
    "datasetBaseUrl": "https://data.labs.kadaster.nl/"
  }

Example URL:  https://labs.kadaster.nl/trifid-kg/kadaster/kg/id/geo-shape/bag-0200100000001088

This contains the following ids in the table heads:
http://labs.kadaster.nl/kadaster/kg/id/geo-shape/bag-0200100000001088
http://labs.kadaster.nl/kadaster/kg/id/place/0200010000090244

@kad-kolka please add when information is missing. 

(FYI; do not look at the links, we added the '/trifid-kg' manually in the function renderLink.)

**Reply by l00mi** [reply]
Okay after some further investigation it seems that the problems might come from your proxy.

You should set the necessary headers correctly to resolve this problem: "x-forwarded-host" / "x-forwarded-proto" .

References:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto

**Reply by kad-kolka** [reply]
I will have look, thanks :-)
